### PR TITLE
fix nan problem for yolov2

### LINKED ISSUE #381: `nan` matrix for all parameters after training for thousands of steps
When training on my own dataset, the parameters and loss may become `nan` suddenly. Does anyone encounter similar problem? What's the exact reason?

### PR DESCRIPTION
**Pull request by chewwt**
resolves #381 

gradients of expit_tensor, tf.pow, and tf.sqrt are unstable when the value is 0.
expit_tensor is replaced with tf.nn.sigmoid
tf.pow is replaced with tf.square
tf.sqrt is fixed according to https://github.com/tensorflow/tensorflow/issues/4914 

Values going into tf.exp is also clipped to min -1e3 and max 10

### CODE REVIEW

File: darkflow/net/yolov2/train.py [hunk 1]
```diff
@@ -56,19 +53,20 @@ def loss(self, net_out):
...
     coords = tf.reshape(coords, [-1, H*W, B, 4])
-    adjusted_coords_xy = expit_tensor(coords[:,:,:,0:2])
-    adjusted_coords_wh = tf.sqrt(tf.exp(coords[:,:,:,2:4]) * np.reshape(anchors, [1, 1, B, 2]) / np.reshape([W, H], [1, 1, 1, 2]))
+    adjusted_coords_xy = tf.nn.sigmoid(coords[:,:,:,0:2])
+    adjusted_coords_wh = tf.exp(tf.clip_by_value(coords[:,:,:,2:4], -1e3, 10)) * np.reshape(anchors, [1, 1, B, 2]) / np.reshape([W, H], [1, 1, 1, 2])
+    adjusted_coords_wh = tf.sqrt(tf.clip_by_value(adjusted_coords_wh, 1e-10, 1e5))
     coords = tf.concat([adjusted_coords_xy, adjusted_coords_wh], 3)
 
-    adjusted_c = expit_tensor(net_out_reshape[:, :, :, :, 4])
+    adjusted_c = tf.nn.sigmoid(net_out_reshape[:, :, :, :, 4])
```
Comment (thtrieu) [inline darkflow/net/yolov2/train.py]: tf.sigmoid is enough

File: darkflow/net/yolov2/train.py [hunk 2]
```diff
@@ -56,19 +53,20 @@ def loss(self, net_out):
     net_out_reshape = tf.reshape(net_out, [-1, H, W, B, (4 + 1 + C)])
     coords = net_out_reshape[:, :, :, :, :4]
     coords = tf.reshape(coords, [-1, H*W, B, 4])
-    adjusted_coords_xy = expit_tensor(coords[:,:,:,0:2])
-    adjusted_coords_wh = tf.sqrt(tf.exp(coords[:,:,:,2:4]) * np.reshape(anchors, [1, 1, B, 2]) / np.reshape([W, H], [1, 1, 1, 2]))
+    adjusted_coords_xy = tf.nn.sigmoid(coords[:,:,:,0:2])
+    adjusted_coords_wh = tf.exp(tf.clip_by_value(coords[:,:,:,2:4], -1e3, 10)) * np.reshape(anchors, [1, 1, B, 2]) / np.reshape([W, H], [1, 1, 1, 2])
```
Comment (thtrieu) [inline darkflow/net/yolov2/train.py]: what are the ranges of coords? why do we clip to, specifically, -1e3 and 10? Source?

Comment (chewwt) [reply darkflow/net/yolov2/train.py]: They were arbitrarily chosen. I had the problem where coords went to inf, so the main purpose is to clip the maximum. I'm actually not too sure how to choose the numbers.

**Comment by sheerun** [comment]
merge please? :)

**Comment by thtrieu** [comment]
Are you sure this one solve #381 ?

**Comment by sheerun** [comment]
I didn't experience this issue since I applied this patch, no issues with training as well

**Comment by onurbarut** [comment]
I tried this and it works! Thanks.
